File: problem/Length3_Palindromic/countPalindromicSubsequence.c
int countPalindromicSubsequence(char *s)
{
    int first[26], last[26];
    for (int i = 0; i < 26; i++)
    {
        first[i] = -1;
        last[i] = -1;
    }

    int n = 0;
    while (s[n] != '\0')
        n++;

    for (int i = 0; i < n; i++)
    {
        int c = s[i] - 'a';
        if (first[c] == -1)
            first[c] = i;
        last[c] = i;
    }

    int ans = 0;

    for (int c = 0; c < 26; c++)
    {
        if (first[c] != -1 && last[c] - first[c] > 1)
        {
            int mask = 0;
            for (int i = first[c] + 1; i < last[c]; i++)
            {
                mask |= 1 << (s[i] - 'a');
            }

            int cnt = 0;
            int x = mask;
            while (x > 0)
            {
                x &= (x - 1);
                cnt++;
            }
            ans += cnt;
        }
    }

    return ans;
}

File: problem/Length3_Palindromic/countPalindromicSubsequence.js
/**
 * @param {string} s
 * @return {number}
 */
var countPalindromicSubsequence = function(s) {
    const n = s.length;
    const first = Array(26).fill(-1);
    const last = Array(26).fill(-1);

    for (let i = 0; i < n; i++) {
        const c = s.charCodeAt(i) - 97;
        if (first[c] === -1) first[c] = i;
        last[c] = i;
    }

    let ans = 0;

    for (let c = 0; c < 26; c++) {
        if (first[c] !== -1 && last[c] - first[c] > 1) {
            let mask = 0;
            for (let i = first[c] + 1; i < last[c]; i++) {
                mask |= 1 << (s.charCodeAt(i) - 97);
            }
            ans += mask.toString(2).split('1').length - 1;
        }
    }

    return ans;
};